SocketRocket/SRWebSocket.m

```diff
@@ -373,6 +373,25 @@ - (void)_HTTPHeadersDidFinish;
 {
     NSInteger responseCode = CFHTTPMessageGetResponseStatusCode(_receivedHTTPHeaders);
     if (responseCode >= 400) {
+        
+        if (responseCode == 401)
+        {
+            // handle an aunathorised request to upgrade
+            if ([self.authenticationDelegate respondsToSelector:@selector(webSocket:bearerAuthenticationDidFailWithRetryHandler:)])
+            {
+                [self.authenticationDelegate webSocket:self bearerAuthenticationDidFailWithRetryHandler:^(NSString *newBearerToken) {
+                    if (newBearerToken)
+                    {
+                        dispatch_async(_workQueue, ^{
+                            self.authenticationBearerToken = newBearerToken;
+                            [self didConnect];
```

Comment:
How do you resolve the infinite loop case when the delegate provides a token but you still receive a 401? If the token provided by the delegate does not resolve the authentication then the original 401 should be propogated to the caller. This is analagous to the HTTP middleware protocol 'fail fast' approach